File: 31-next-permutation/31-next-permutation.scala
object Solution {
    def nextPermutation(nums: Array[Int]): Unit = {
        var i: Int = nums.length - 1
        var found = false

        while(i > 0 && !found) {
            if (nums(i) > nums(i - 1)) found = true
            else i -= 1
        }

        i -= 1
        
        if (i == -1) {
            reverse(nums, 0, nums.length - 1)
        } else {
            var k = i + 1
            var j = i + 1

            while(k < nums.length) {
                if (nums(k) > nums(i)) j = k
                k += 1
            }

            swap(nums, i, j)
            reverse(nums, i + 1, nums.length - 1)
        }
    }

    def swap(nums: Array[Int], a: Int, b: Int): Unit = {
        val tmp = nums(a)
        nums(a) = nums(b)
        nums(b) = tmp
    }

    def reverse(nums: Array[Int], start: Int, end: Int): Unit = {
        var i = start
        var j = end
        while(i < j) {
            swap(nums, i, j)
            i += 1
            j -= 1
        }
    }
}

File: 31-next-permutation/31-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        int p = nums[nums.length - 1];
        int i = nums.length - 2;
        while(i >= 0) {
            if (nums[i] < nums[i + 1]) break;
            i--;
        }
        
        if (i == -1) {
            reverse(nums, 0, nums.length - 1);
            return;
        }
        
        int j = nums.length - 1;
        while(nums[j] <= nums[i]) j--;
        
        swap(nums, i, j);
        reverse(nums, i + 1, nums.length - 1);
    }
    
    void swap(int[] nums, int a, int b) {
        int tmp = nums[a];
        nums[a] = nums[b];
        nums[b] = tmp;
    }
    
    void reverse(int[] nums, int s, int e) {
        while(s < e) {
            swap(nums, s++, e--);
        }
    }
}